/*
 * fam_metadata_rpc.proto
 * Copyright (c) 2020, 2023 Hewlett Packard Enterprise Development, LP. All
 * rights reserved. Redistribution and use in source and binary forms, with or
 * without modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * 3. Neither the name of the copyright holder nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission.
 *
 *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
 * IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * See https://spdx.org/licenses/BSD-3-Clause
 *
 */

syntax = "proto3";

/*
 *define services for RPC methods
 */
service Fam_Metadata_Rpc {
    rpc signal_start(Fam_Metadata_Gen_Request)
        returns (Fam_Metadata_Gen_Response) {}
    rpc signal_termination(Fam_Metadata_Gen_Request)
        returns (Fam_Metadata_Gen_Response) {}
    rpc metadata_insert_region(Fam_Metadata_Region_Request)
        returns (Fam_Metadata_Response) {}
    rpc metadata_delete_region(Fam_Metadata_Request)
        returns (Fam_Metadata_Response) {}
    rpc metadata_find_region(Fam_Metadata_Request)
        returns (Fam_Metadata_Region_Response) {}
    rpc metadata_modify_region(Fam_Metadata_Region_Request)
        returns (Fam_Metadata_Response) {}
    rpc metadata_insert_dataitem(Fam_Metadata_Request)
        returns (Fam_Metadata_Response) {}
    rpc metadata_delete_dataitem(Fam_Metadata_Request)
        returns (Fam_Metadata_Response) {}
    rpc metadata_find_dataitem(Fam_Metadata_Request)
        returns (Fam_Metadata_Response) {}
    rpc metadata_modify_dataitem(Fam_Metadata_Request)
        returns (Fam_Metadata_Response) {}
    rpc metadata_check_region_permissions(Fam_Permission_Request)
        returns (Fam_Permission_Response) {}
    rpc metadata_check_item_permissions(Fam_Permission_Request)
        returns (Fam_Permission_Response) {}
    rpc metadata_maxkeylen(Fam_Metadata_Request)
        returns (Fam_Metadata_Response) {}
    rpc reset_profile(Fam_Metadata_Gen_Request)
        returns (Fam_Metadata_Gen_Response) {}
    rpc dump_profile(Fam_Metadata_Gen_Request)
        returns (Fam_Metadata_Gen_Response) {}
    rpc metadata_update_memoryserver(Fam_Memservcnt_Request)
        returns (Fam_Metadata_Gen_Response) {}
    rpc metadata_reset_bitmap(Fam_Metadata_Request)
        returns (Fam_Metadata_Gen_Response) {}
    rpc metadata_validate_and_create_region(Fam_Metadata_Request)
        returns (Fam_Metadata_Region_Info_Response) {}
    rpc metadata_validate_and_destroy_region(Fam_Metadata_Request)
        returns (Fam_Metadata_Region_Info_Response) {}
    rpc metadata_validate_and_allocate_dataitem(Fam_Metadata_Request)
        returns (Fam_Metadata_Response) {}
    rpc metadata_validate_and_deallocate_dataitem(Fam_Metadata_Request)
        returns (Fam_Metadata_Response) {}
    rpc metadata_find_region_and_check_permissions(Fam_Metadata_Region_Request)
       returns (Fam_Metadata_Region_Response) {}
    rpc metadata_find_dataitem_and_check_permissions(Fam_Metadata_Request)
       returns (Fam_Metadata_Response) {}
    rpc get_memory_server_list(Fam_Metadata_Request)
        returns (Fam_Metadata_Region_Info_Response) {}
}

/*
 * Request message used by methods signal_start and signal_termination
 */
message Fam_Metadata_Gen_Request {}

message Fam_Metadata_Region_Request {
    oneof metadata_key {
        uint64 key_region_id = 1;
        uint64 key_dataitem_id = 2;
    }
    string key_region_name = 3;
    string key_dataitem_name = 4;
    uint64 region_id = 5;
    string name = 6;
    uint64 offset = 7;
    uint32 uid = 8;
    uint32 gid = 9;
    uint64 perm = 10;
    uint64 size = 11;
    uint32 user_policy = 12;
    uint64 memsrv_cnt = 13;
    repeated uint64 memsrv_list = 14;
    uint64 memsrv_id = 15;
    uint32 op = 16;
    uint32 redundancylevel = 17;
    uint32 memorytype = 18;
    uint32 interleaveenable = 19;
    uint64 interleavesize = 20;
    uint32 permission_level = 21;
    uint32 allocpolicy = 22; 
    uint32 hostingnode = 23;
}

message Fam_Metadata_Region_Response {
    uint64 region_id = 1;
    string name = 2;
    uint64 offset = 3;
    uint32 uid = 4;
    uint32 gid = 5;
    uint64 perm = 6;
    uint64 size = 7;
    uint64 maxkeylen = 8;
    bool isfound = 9;
    uint32 errorcode = 10;
    string errormsg = 11;
    uint64 memsrv_id = 12;
    repeated uint64 memsrv_list = 13;
    uint64 memsrv_cnt = 14;
    uint32 redundancylevel = 15;
    uint32 memorytype = 16;
    uint32 interleaveenable = 17;
    uint64 interleavesize = 18;
    uint32 permission_level = 19;
    uint32 allocpolicy = 20;
}
/*
 * Response message used by methods signal_start and signal_termination
 */
message Fam_Metadata_Gen_Response {
    repeated fixed32 addrname = 1;
    uint64 addrnamelen = 2;
}

message Fam_Metadata_Request {
    oneof metadata_region_key {
        uint64 key_region_id = 1;
        string key_region_name = 2;
    }
    oneof metadata_dataitem_key {
        uint64 key_dataitem_id = 3;
        string key_dataitem_name = 4;
    }
    uint64 region_id = 5;
    uint64 dataitem_id = 6;
    string name = 7;
    repeated uint64 offsets = 8;
    uint32 uid = 9;
    uint32 gid = 10;
    uint64 perm = 11;
    uint64 size = 12;
    uint32 user_policy = 13;
    uint64 memsrv_id = 14;
    uint32 op = 15;
    uint32 redundancylevel = 16;
    uint32 memorytype = 17;
    uint32 interleaveenable = 18;
    uint64 interleave_size = 19;
    repeated uint64 memsrv_list = 20;
    uint64 memsrv_cnt = 21;
    uint64 permission_level = 22;
    uint32 allocpolicy = 23; 
    uint32 hostingnode = 24;
}

message Fam_Metadata_Response {
    uint64 region_id = 1;
    string name = 2;
    repeated uint64 offsets = 3;
    uint32 uid = 4;
    uint32 gid = 5;
    uint64 perm = 6;
    uint64 size = 7;
    uint64 maxkeylen = 8;
    bool isfound = 9;
    uint32 errorcode = 10;
    string errormsg = 11;
    uint64 memsrv_id = 12;
    uint64 interleave_size = 13;
    repeated uint64 memsrv_list = 14;
    uint64 memsrv_cnt = 15;
    uint64 permission_level = 16;
    uint64 region_permission = 17;
    repeated fixed32 addrname = 18;
    uint64 addrnamelen = 19;
}

message Fam_Permission_Request {
    enum meta_ops {
        META_REGION_ITEM_WRITE = 0;
        META_REGION_ITEM_READ = 1;
        META_REGION_ITEM_EXEC = 2;
        META_REGION_ITEM_RW = 3;
    }
    uint32 uid_meta = 1;
    uint32 gid_meta = 2;
    uint64 perm = 3;
    uint32 uid_in = 4;
    uint32 gid_in = 5;
    uint32 ops = 6;
}

message Fam_Permission_Response {
    bool is_permitted = 1;
    uint32 errorcode = 2;
    string errormsg = 3;
}

message Fam_Memservcnt_Request {
    uint32 nmemserverspersistent = 1;
    repeated uint64 memsrv_persistent_list = 2;
    uint32 nmemserversvolatile = 3;
    repeated uint64 memsrv_volatile_list = 4;
}

message Fam_Metadata_Region_Info_Response {
    uint64 region_id = 1;
    repeated uint32 memsrv_list = 2;
    uint32 errorcode = 3;
    string errormsg = 4;
    uint64 memsrv_id = 5;
}
